### PR TITLE
Add table aws_ec2_reserved_instance #599

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_ec2_reserved_instance []

PRETEST: tests/aws_ec2_reserved_instance

TEST: tests/aws_ec2_reserved_instance
Running terraform
^[[Ddata.aws_region.alternate: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
null_resource.named_test_resource: Creating...
null_resource.named_test_resource: Provisioning with 'local-exec'...
null_resource.named_test_resource (local-exec): Executing: ["/bin/sh" "-c" "aws ec2 describe-reserved-instances --output json > /private/var/folders/5q/jqv70bxs7g73gwgv5tvl80l40000gn/T/tests/aws_ec2_reserved_instance/terraform/test/output.json"]
null_resource.named_test_resource: Creation complete after 2s [id=3375260713042356500]
data.local_file.input: Refreshing state...

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

currency_code = USD
fixed_price = 0
instance_type = t3.nano
offering_class = standard
resource_aka = arn:aws:ec2:us-east-1:632902152528:reserved-instances/0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b
resource_name = 0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b
scope = Region
state = active
usage_price = 0

Running SQL query: test-get-query.sql
[
  {
    "arn": "arn:aws:ec2:us-east-1:632902152528:reserved-instances/0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b",
    "currency_code": "USD",
    "fixed_price": "0",
    "instance_state": "active",
    "instance_type": "t3.nano",
    "offering_class": "standard",
    "reserved_instance_id": "0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b",
    "scope": "Region",
    "usage_price": "0"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "arn": "arn:aws:ec2:us-east-1:632902152528:reserved-instances/0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b",
    "instance_type": "t3.nano",
    "reserved_instance_id": "0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:ec2:us-east-1:632902152528:reserved-instances/0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b"
    ],
    "title": "0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b"
  }
]
✔ PASSED

POSTTEST: tests/aws_ec2_reserved_instance

TEARDOWN: tests/aws_ec2_reserved_instance

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic Info

```sql
select 
  reserved_instance_id, 
  arn, 
  instance_type, 
  instance_state, 
  currency_code, 
  CAST(fixed_price AS varchar), 
  offering_class, scope, 
  CAST(usage_price AS varchar)
from 
  aws.aws_ec2_reserved_instance;
```
```
+--------------------------------------+----------------------------------------------------------------------------------+---------------+----------------+---------------+-------------+----------------+--------+-------------+
| reserved_instance_id                 | arn                                                                              | instance_type | instance_state | currency_code | fixed_price | offering_class | scope  | usage_price |
+--------------------------------------+----------------------------------------------------------------------------------+---------------+----------------+---------------+-------------+----------------+--------+-------------+
| 0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b | arn:aws:ec2:us-east-1:632902152528:reserved-instances/0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b | t3.nano       | active         | USD           | 0           | standard       | Region | 0           |
+--------------------------------------+----------------------------------------------------------------------------------+---------------+----------------+---------------+-------------+----------------+--------+-------------+
```

### Count the number of instances by instance type

```sql
select
  instance_type,
  count(instance_count) as count
from
  aws_ec2_reserved_instance
group by
  instance_type;
```
```
+---------------+-------+
| instance_type | count |
+---------------+-------+
| t3.nano       | 1     |
+---------------+-------+
```

### List reserved instances of standard offering class

```sql
select
  reserved_instance_id,
  instance_type,
  offering_class
from
  aws_ec2_reserved_instance
where
  offering_class = 'standard';
```

```
+--------------------------------------+---------------+----------------+
| reserved_instance_id                 | instance_type | offering_class |
+--------------------------------------+---------------+----------------+
| 0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b | t3.nano       | standard       |
+--------------------------------------+---------------+----------------+
```

### List active reserved instances

```sql
select 
  reserved_instance_id,
  instance_type,
  instance_state
from 
  aws_ec2_reserved_instance
where
  instance_state = 'active';
```

```
+--------------------------------------+---------------+----------------+
| reserved_instance_id                 | instance_type | instance_state |
+--------------------------------------+---------------+----------------+
| 0bdff5fd-4b22-470c-8e1d-3eaff8a4ec3b | t3.nano       | active         |
+--------------------------------------+---------------+----------------+

```

</details>

### CODE REVIEW

File: aws/table_aws_ec2_reserved_instance.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	return &plugin.Table{
+		Name:        "aws_ec2_reserved_instance",
+		Description: "AWS EC2 Reserved Instance",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.SingleColumn("reserved_instance_id"),
+			ShouldIgnoreError: isNotFoundError([]string{"InvalidParameterValue", "InvalidInstanceID.Unavailable", "InvalidInstanceID.Malformed"}),
+			Hydrate:           getEc2ReservedInstance,
+		},
+		List: &plugin.ListConfig{
+			Hydrate: listEc2ReservedInstance,
```
Comment (khushboo9024) [inline aws/table_aws_ec2_reserved_instance.go:26]:
```suggestion
			Hydrate: listEc2ReservedInstances,
```

File: aws/table_aws_ec2_reserved_instance.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listEc2ReservedInstance(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	region := d.KeyColumnQualString(matrixKeyRegion)
+	plugin.Logger(ctx).Trace("listEc2Instance", "AWS_REGION", region)
```
Comment (khushboo9024) [inline aws/table_aws_ec2_reserved_instance.go:163]:
```suggestion
	plugin.Logger(ctx).Trace("listEc2ReservedInstances", "AWS_REGION", region)
```

File: docs/tables/aws_ec2_reserved_instance.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  count(*) as count
+from
+  aws_ec2_reserved_instance
+where
+  instance_type not in ('t2.large', 'm3.medium')
+group by
+  instance_type;
+```
+
+### List reserved instances of standard offering class
```
Comment (khushboo9024) [inline docs/tables/aws_ec2_reserved_instance.md:49]:
```suggestion
### List standard offering class type reserved instances
```

File: docs/tables/aws_ec2_reserved_instance.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+select
+  instance_type,
+  count(instance_count) as count
+from
+  aws_ec2_reserved_instance
+group by
+  instance_type;
+```
+
+### List of EC2 reserved instances provisioned with undesired(for example t2.large and m3.medium is desired) instance type(s).
```
Comment (khushboo9024) [inline docs/tables/aws_ec2_reserved_instance.md:35]:
```suggestion
### List reserved instances provisioned with undesired(for example t2.large and m3.medium is desired) instance type(s)
```

File: docs/tables/aws_ec2_reserved_instance.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  instance_state, 
+  currency_code, 
+  CAST(fixed_price AS varchar), 
+  offering_class, scope, 
+  CAST(usage_price AS varchar)
+from 
+  aws.aws_ec2_reserved_instance;
+```
+
+### Count the number of instances by instance type
```
Comment (khushboo9024) [inline docs/tables/aws_ec2_reserved_instance.md:23]:
```suggestion
### Count reserved instances by instance type
```

File: docs/tables/aws_ec2_reserved_instance.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  reserved_instance_id, 
+  arn, 
+  instance_type, 
+  instance_state, 
+  currency_code, 
+  CAST(fixed_price AS varchar), 
+  offering_class, scope, 
+  CAST(usage_price AS varchar)
+from 
+  aws.aws_ec2_reserved_instance;
```
Comment (khushboo9024) [inline docs/tables/aws_ec2_reserved_instance.md:20]:
```suggestion
 aws_ec2_reserved_instance;
```

File: aws/table_aws_ec2_reserved_instance.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+	region := d.KeyColumnQualString(matrixKeyRegion)
+
+	getCommonColumnsCached := plugin.HydrateFunc(getCommonColumns).WithCache()
+	commonData, err := getCommonColumnsCached(ctx, d, h)
+	if err != nil {
+		return nil, err
+	}
+	commonColumnData := commonData.(*awsCommonColumnData)
+
+	arn := "arn:" + commonColumnData.Partition + ":ec2:" + region + ":" + commonColumnData.AccountId + ":instance/" + *instance.ReservedInstancesId
```
Comment (khushboo9024) [inline aws/table_aws_ec2_reserved_instance.go:225]: Arn should be `arn:${Partition}:ec2:${Region}:${Account}:reserved-instances/${ReservationId}`, please check?

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Pull request merged**